feat: Auto height instant update

riodeuno opened the pull request:
## Description
This PR adds one of the promised updates to the auto height feature. 
More specifically, we wanted to add was the ability to see the containers change height as we drag and drop widgets within them instead of after dropping (when auto height is enabled)
This PR does that.

- [x] Display the auto height updates during drag and resize, instead of after drop. _(see video below)_

- [ ] Include scrollbars within borders of the container like widgets such that the borders are not overridden by scrollbars and auto height considers border thickness for computations

- [x] Make Container widgets leaner by reducing the number of DOM nodes which make up a container like widget. _This reduces the load on the browser, so that it doesn't have to render as many DOM nodes as before. So, this is improving rendering performance, however, it may not be noticeable, as it is in milliseconds. We haven't registered any metrics for this, as this falls in the best practices category instead of the performance improvement category._


Fixes #19216 
Fixes #19578 
Fixes #19579 


Media

https://user-images.githubusercontent.com/103687/208754409-dc2a7058-7fb2-477a-a1c4-562ab3b120c9.mov



## Type of change

- This changes the way auto height applies updates, particularly w.r.t borders in container like widgets.
- This change will result in a different experience, which is an improvement on the existing experience.


## How Has This Been Tested?
- Manual: Scenarios will be tested manually. Please see Test Plan for more information.
- Jest: Functions added have unit tests
This involves drag and drop and will not have cypress tests.

### Test Plan
https://github.com/appsmithorg/TestSmith/issues/2145

### Issues raised during DP testing
https://github.com/appsmithorg/appsmith/pull/19082#issuecomment-1401946649
https://github.com/appsmithorg/appsmith/pull/19082#issuecomment-1402369139
https://github.com/appsmithorg/appsmith/pull/19082#issuecomment-1406532583

## Checklist:
### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [NA] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [NA] PR is being merged under a feature flag


### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

SatishGandham commented:
/perf-test
Testing performance infra, ignore.

riodeuno commented on app/client/src/actions/autoHeightActions.ts:
```diff
@@ -22,12 +23,14 @@ export function setAutoHeightLayoutTreeAction(
     payload: {
       shouldCheckContainersForAutoHeightUpdates,
       layoutUpdated: !!layoutUpdated,
+      resettingTabs: !!resettingTabs,
```
This will be removed if we don't find any issues with the Tabs Widget.

riodeuno commented on app/client/src/components/editorComponents/DraggableComponent.tsx:
```diff
@@ -120,17 +104,6 @@ function DraggableComponent(props: DraggableComponentProps) {
   const dragWrapperStyle: CSSProperties = {
     display: isCurrentWidgetDragging ? "none" : "block",
   };
-  const dragBoundariesStyle: React.CSSProperties = useMemo(() => {
```
@rahulramesha @marks0351 I'll need some information on what this was doing. I've removed it, because, it seemed like it wasn't helpful.

riodeuno replied:
Nevermind, I got what this is about. I'll try to put this back.

riodeuno commented on app/client/src/sagas/autoHeightSagas/containers.ts:
```diff
@@ -18,7 +21,14 @@ import { getChildOfContainerLikeWidget } from "./helpers";
 
-export function* dynamicallyUpdateContainersSaga() {
+export function* dynamicallyUpdateContainersSaga(
+  // Added this aciton payload to handle a scenario where the tabs widget's default is not Tab 1.
```
A comment to remember to do this.

riodeuno commented on app/client/src/utils/autoHeight/mutateDOM.ts:
```diff
@@ -0,0 +1,116 @@
+// Here the data structure is the `widgetsToUpdate` data structure. If possible, we should create the `updates`
+// we use in the function directly in the `widgets.ts` (auto height saga)
+// This way, we can avoid looping again in this place.
+function getNodesAndStylesToUpdate(
```
Add unit tests if possible.

riodeuno commented on app/client/src/utils/WidgetPropsUtils.tsx:
```diff
@@ -339,3 +343,105 @@ export const generateWidgetProps = (
+  return offset;
+};
+
+export function getCanvasWidgetHeightsToUpdate(
```
Add function doc and unit tests.

riodeuno commented on app/client/src/utils/WidgetPropsUtils.tsx:
```diff
@@ -339,3 +343,105 @@ export const generateWidgetProps = (
+  return updatedCanvasWidgets;
+}
+
+export function getCanvasBottomRow(
```
Add function doc and tests

riodeuno commented:
@rahulramesha @marks0351 @kamakshibhat-appsmith @Sripriya93 
This is now ready for QA and review.

riodeuno commented:
/ok-to-test sha=093649e

riodeuno commented on app/client/src/pages/Editor/MainContainer.test.tsx:
```diff
@@ -562,8 +562,10 @@ describe("Drag and Drop widgets into Main container", () => {
     updatedLength = updatedDropTarget.style.height;
-    expect(getAbsolutePixels(initialLength) + amountMovedY).toEqual(
-      getAbsolutePixels(updatedLength),
+    // the above update to the main container now happens via a separate pathway (mutateDOM), and is not updated in the component
```
@marks0351 Please check this out.
In the meantime, I'll check with @appsmithorg/qa to see if we can automate this in cypress tests.

riodeuno commented:
/ok-to-test sha=bbaa131

Sripriya93 commented:
Observations @riodeuno 

1. Container as parent and tabs widget inside it, and when we drag a checkbox to go to the bottom tabs and container are both not resizing

https://user-images.githubusercontent.com/104895602/214274093-4c713200-eb89-4684-80b8-1978c26d7e4c.mp4

2. Tabs widget when we resize any widget inside it, we are seeing the dotted border of the tabs to be slightly above the actual size of the tabs

https://user-images.githubusercontent.com/104895602/214275855-38be1515-74a1-4224-bc0e-135e6b81a854.mp4

3. Canvas has a gap at the bottom when we drag and drop any widget in it.

![image](https://user-images.githubusercontent.com/104895602/214279800-b3697d35-f3e6-4f77-aae0-6aeebb95dba9.png)
 
4. Container as parent and tabs widget inside it, and when we drag a smaller widget to the bottom, we are able to see the scroll goes up a bit and reduces in height and when we drop the height enlarges to fit the widget.

https://user-images.githubusercontent.com/104895602/214284257-3607d8ab-6993-45e5-b662-4988d7a537a3.mp4

riodeuno commented:
@Sripriya93 Thank you for spotting these issues, I'll push fixes for this by tomorrow.

Sripriya93 commented:
5. Map widget have issue with displaying the map itself

![image](https://user-images.githubusercontent.com/104895602/214362993-e9bba502-4831-472d-91e5-60db454e3f24.png)

6. Able to see a huge space below Modal widget when we drag and drop it

https://user-images.githubusercontent.com/104895602/214363663-8e48648b-0fcf-4090-8887-62236d24e4ea.mp4

7. Modal widget collapses when we DnD a Tab widget inside it. (Also happens with JSON form, Switch group & stat box (When we move it a bit after placing it inside the Modal), )

https://user-images.githubusercontent.com/104895602/214364624-f9fbc640-e36c-4268-84b0-c4b70528088f.mp4

8. For Modal widget, we are seeing a scroll overriding the border of the widget.

<img width="290" alt="image" src="https://user-images.githubusercontent.com/104895602/214366164-0bd2ad30-feae-48b7-a8e1-50c77a6f627f.png">

9. Stat box inside Modal widget when we try to move it inside the Modal we are not able to see the position or the highlight of where the widget is placed at that moment 

https://user-images.githubusercontent.com/104895602/214368362-329ca407-d2bf-430d-9221-aac1271f9eea.mp4

10. Tabs widget, unnecessarily gets expand when we try to drop any widget inside the child widget, and then returning to the original height after the drop

https://user-images.githubusercontent.com/104895602/214370642-4e75f0ff-3360-408f-8c61-5287c3bdc6f7.mp4

11. Issue with Camera widget, size of the screen inside the camera is small

<img width="328" alt="image" src="https://user-images.githubusercontent.com/104895602/214371477-c3165bbe-f238-40c7-9510-c4aff657947d.png">

12. Progress widget is blank

<img width="272" alt="image" src="https://user-images.githubusercontent.com/104895602/214372205-ef8507d8-e5a4-4c1c-ac0b-728e5f6d8916.png">

marks0351 commented on app/client/src/widgets/ContainerWidget/component/index.tsx:
```diff
@@ -1,123 +1,132 @@
-        `
-      : ""}
-
+  ${(props) => (props.noScroll ? `` : scrollCSS)}
```
pls check if box-shadow is necessary.

riodeuno replied:
Doesn't seem necessary, as it uses props.selected. which doesn't make sense.

riodeuno commented on app/client/src/components/editorComponents/ResizableComponent.tsx:
```diff
@@ -286,28 +286,35 @@ export const ResizableComponent = memo(function ResizableComponent(
   }, [props, isAutoHeightEnabledForWidget, isEnabled]);
 
+  const fixedHeight =
+    isAutoHeightEnabledForWidget(props, true) ||
```
Created an issue #20110

riodeuno commented on app/client/src/components/designSystems/appsmith/WidgetStyleContainer.tsx:
```diff
@@ -35,19 +35,19 @@ const WidgetStyle = styled.div<WidgetStyleContainerProps>`
   background-color: ${(props) => props.backgroundColor || "transparent"};
+  overflow: hidden;
 
   & > div {
```
Doesn't seem necessary.

marks0351 replied:
looks like its still exists.

riodeuno replied:
Yes, I forgot to mention it here. This turns out to be necessary. I found this when I was fixing QA issues.

riodeuno replied:
I added it back in this commit: https://github.com/appsmithorg/appsmith/pull/19082/commits/0d9417271083feb46f811dc3e85b2b381458cb8a

riodeuno commented on app/client/src/components/editorComponents/ResizableComponent.tsx:
```diff
@@ -286,28 +286,35 @@ export const ResizableComponent = memo(function ResizableComponent(
       zWidgetType={props.type}
     >
       <VisibilityContainer
+        className={`visibility-container-${props.widgetId}`}
```
Removed

riodeuno commented on app/client/src/components/editorComponents/DropTargetComponent.tsx:
```diff
@@ -69,39 +68,112 @@ export const DropTargetContext: Context<{
+      // based on the auto height computations
+      // This also updates any "dropTargets" that need to change height
+      // hence, this and the `updateHeight` function are mutually exclusive.
+      if (isParentAutoHeightEnabled || widgetId === MAIN_CONTAINER_WIDGET_ID) {
```
Done.

riodeuno commented on app/client/src/actions/autoHeightActions.ts:
```diff
@@ -45,8 +48,24 @@ export function updateWidgetAutoHeightAction(
+  };
+}
+
+export function immediatelyUpdateAutoHeightAction(
```
Used DOM keyword in the name.

riodeuno commented on app/client/src/utils/WidgetPropsUtils.tsx:
```diff
@@ -339,3 +334,118 @@ export const generateWidgetProps = (
+              childCanvasWidgetId,
+              canvasWidgets,
+            );
+            if (bottomRow > 0)
```
Done.

riodeuno commented on app/client/src/sagas/autoHeightSagas/helpers.ts:
```diff
@@ -162,5 +163,8 @@ export function* getMinHeightBasedOnChildren(
     }
   }
 
+  if (widgetId === MAIN_CONTAINER_WIDGET_ID)
```
Done

riodeuno commented on app/client/src/pages/common/CanvasArenas/StickyCanvasArena.tsx:
```diff
@@ -123,7 +123,8 @@ export const StickyCanvasArena = forwardRef(
     const observeSlider = () => {
       interSectionObserver.current.disconnect();
-      interSectionObserver.current.observe(slidingArenaRef.current);
+      if (slidingArenaRef && slidingArenaRef.current)
```
Done.

riodeuno commented on app/client/src/components/editorComponents/ErrorBoundry.tsx:
```diff
@@ -42,7 +34,7 @@ class ErrorBoundary extends React.Component<Props, State> {
   render() {
     return (
-      <ErrorBoundaryContainer>
+      <ErrorBoundaryContainer className="error-boundary">
```
Done. Seems to be working fine.

riodeuno commented on app/client/src/reducers/entityReducers/canvasWidgetsReducer.ts:
```diff
@@ -101,6 +123,15 @@ const canvasWidgetsReducer = createImmerReducer(initialState, {
       });
     }
+
+    const canvasWidgetHeightsToUpdate: Record<
```
As discussed. We will make a conscious choice to make this change. I'll create a performance product note and share it on the team performance channel.

riodeuno commented on app/client/src/components/designSystems/appsmith/WidgetStyleContainer.tsx:
```diff
@@ -35,19 +35,19 @@ const WidgetStyle = styled.div<WidgetStyleContainerProps>`
 function WidgetStyleContainer(props: WidgetStyleContainerProps) {
   return (
     <WidgetStyle {...props} data-testid={`container-wrapper-${props.widgetId}`}>
-      <div>{props.children}</div>
```
@Sripriya93 @kamakshibhat-appsmith Please help me test this.

riodeuno commented:
@marks0351 @rahulramesha This is once again ready for review.

riodeuno commented:
@kamakshibhat-appsmith @Sripriya93 This is ready for QA with the fixes for the known issues.

riodeuno commented:
/ok-to-test sha=2243f98

Sripriya93 commented:
@riodeuno here are the issues found today
1. Large space below the Modal widget is still seen (Already discussed please ignore)

2. White space seen below canvas, Is it expected?
![image](https://user-images.githubusercontent.com/104895602/215020833-3887a833-3cdd-403a-8ef0-03d1273c59bd.png)

3. Tabs widget, unnecessarily gets expand when we try to drop any widget inside the child widget, and then returning to the expected height after the drop, because of this when we put tabs widget inside container, we are not able to drop any widget at the bottom of the container, even if we scroll

4. When we have Tabs widget inside container and drop a widget inside Tab and scroll down in the container to place a widget at the bottom of the container, and move the mouse the tabs and container scroll back 

https://user-images.githubusercontent.com/104895602/215035259-d7a80d30-7e16-4220-8878-9a43ad31f077.mp4

5. Slider are all Top aligned, in release they are center aligned

6. No signifiers seen in Modal widget

https://user-images.githubusercontent.com/104895602/215056189-1066cd61-fd8b-4068-be45-4f545f14cf1c.mp4

7. Scroll contents toggle not working for Forms widget, in Fixed height

<img width="743" alt="image" src="https://user-images.githubusercontent.com/104895602/215077418-e907b5fb-6287-4eaf-9768-bcd4d2fd673d.png">

8. When we have widgets inside Tabs in Fixed height and when toggle off Scroll contents, we are seeing the widgets inside them disappears

https://user-images.githubusercontent.com/104895602/215076893-c58972ea-85d3-4d26-80b9-5da860d95131.mp4

9. Scroll doesnt get smaller when we remove from the bottom of the canvas

10. Widgets at the corner of the Container and Form widget with Border radius as 1.5 rem, widgets at the corner are not respecting the border radius

<img width="497" alt="image" src="https://user-images.githubusercontent.com/104895602/215094637-3ae96186-5bac-4743-9780-91053c89d4da.png">
<img width="334" alt="image" src="https://user-images.githubusercontent.com/104895602/215094742-f9888ebb-1617-415c-a9d4-381474f390d8.png">

11. In all container like widgets, in fixed height when resized to larger size after we drop two or three widgets, we arent able to drop another widget at the bottom of the widget

https://user-images.githubusercontent.com/104895602/215097699-9f8b2c4b-363e-4261-8fed-cf2ab9cb5c01.mp4

12. In fixed height in Modal widget scroll isnt working, so we are not able to place widgets

https://user-images.githubusercontent.com/104895602/215098588-60beff7e-9128-4c1f-a25f-37440b16d137.mp4

riodeuno commented:
@Sripriya93 First of all, thanks for finding these issues. These are issues I would not have noticed, and thanks to you finding them, I have more confidence in this feature.

> White space seen below canvas, Is it expected?
I need help replicating this issue. Please show me the steps to do this.

> Tabs widget, unnecessarily gets expand when we try to drop any widget inside the child widget, and then returning to the expected height after the drop, because of this when we put tabs widget inside container, we are not able to drop any widget at the bottom of the container, even if we scroll

This is a direct consequence of auto height instant update. I will not be fixing this in this release. We need to go back to the drawing board to figure out a good approach here.

> When we have Tabs widget inside container and drop a widget inside Tab and scroll down in the container to place a widget at the bottom of the container, and move the mouse the tabs and container scroll back

This happens because on focus, the widget name component shows up. This adds a small scroll.

> Slider are all Top aligned, in release they are center aligned

Fixed

> No signifiers seen in Modal widget

Fixed

> Scroll contents toggle not working for Forms widget, in Fixed height

Fixed

> When we have widgets inside Tabs in Fixed height and when toggle off Scroll contents, we are seeing the widgets inside them disappears

Fixed

> Widgets at the corner of the Container and Form widget with Border radius as 1.5 rem, widgets at the corner are not respecting the border radius

Fixed

> In all container like widgets, in fixed height when resized to larger size after we drop two or three widgets, we arent able to drop another widget at the bottom of the widget

I can't seem to replicate this. I believe the fixes for the above issues, must have fixed this, please let me know if this is consistently happening

> In fixed height in Modal widget scroll isnt working, so we are not able to place widgets

Fixed

riodeuno commented:
/ok-to-test sha=4e7c4c6

Sripriya93 commented:
@riodeuno Have retested the fixed issues and saw some issues mentioned below, will take a look at the PR again later today or tomorrow, 
1. Container like widgets are not reflowing (At the beginning of the loom we would be able to see the white space after we DnD a container widget)

https://user-images.githubusercontent.com/104895602/215408669-cd4f68c7-6c6c-4cb5-8b26-16b853410db1.mp4

2. Widget jumps when we drag and drop while reflow 

https://user-images.githubusercontent.com/104895602/215409520-a8574e59-2bd8-4e85-8dd0-754db54b808f.mp4

riodeuno commented:
/ok-to-test sha=2e6dd04

marks0351 commented:
PR looks good. let me know once all bugs are fixed, il approve it one last time.

riodeuno commented:
@Sripriya93 
> Container like widgets are not reflowing

I noticed the same experience on release, so I will not be fixing this in this PR.

> Widget jumps when we drag and drop while reflow

This is a consequence of the widget needing more height when the auto height is computed. This is also something which we won't be fixing in this PR. This needs a product discussion.

riodeuno commented:
/ok-to-test sha=3890aa8

riodeuno commented:
/ok-to-test sha=769ea75

Sripriya93 commented:
@riodeuno This PR is good to go.

riodeuno commented:
/ok-to-test sha=433b06f

riodeuno commented:
/ok-to-test sha=433b06f

riodeuno commented:
/ok-to-test sha=f7e5f91

riodeuno commented:
/perf-test

kamakshibhat-appsmith commented:
With this PR , do we expect fix for the issue where scrollbars were coming in deploy mode when border width was increased for container type(auto height) widgets ? @riodeuno

riodeuno commented:
@kamakshibhat-appsmith Yes, this PR should have fixed that. I can replicate this on release right now. Let me look into this.
